### ISSUE TITLE
PyTorch Equivalent of Numpy's nanmean (or add exclude_nans to mean/var/other reductions)

### ISSUE THREAD

**Issue by RylanSchaeffer** [pytorch/pytorch#21987]
## 🚀 Feature
Numpy has a function, `np.nanmean()`, that excludes NaN values when computing the mean. I'd like 

## Motivation
Suppose I want to compute MSE over two vectors, one of which has NaN values. Right now, there's no easy way of doing this.

## Hope
`torch.nanmean(x)` returns the mean of all non-nan values of `x`.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @mruberry @rgommers @heitorschueroff

**Comment by umanwizard** [comment]
I think instead of reproducing the NumPy interface exactly, it would be better if we added a parameter `exclude_nans` to `mean`.

That way, we wouldn't have to add a million new overloads (there are already at least 9 overloads for `mean`).

**Comment by RylanSchaeffer** [comment]
@umanwizard sounds good to me!

**Comment by SsnL** [comment]
ohh the np nan* methods! I certain do wish we have a way to do them in pytorch too!

**Comment by apaszke** [comment]
:+1: for making this a flag

**Comment by ezyang** [comment]
@umanwizard are you going to implement this? :) We think that this is a good idea./

**Comment by umanwizard** [comment]
Yes i am

**Comment by RylanSchaeffer** [comment]
In the interim, do you have any recommendations for how to exclude `NaN` while computing the batch-wise mean or sum over a `pack_padded_sequence`?

**Comment by RylanSchaeffer** [comment]
For clarify, suppose I have some tensor of shape `(batch size, max sequence length, output dimension)` and many of the last few elements of the sequences are `NaN`. If I want to reduce this down to the batch-wise mean of shape `(batch size)`, how do I do this?

**Comment by yulkang** [comment]
Doesn't this do the job:
```python
def nanmean(v, *args, inplace=False, **kwargs):
    if not inplace:
        v = v.clone()
    is_nan = isnan(v)
    v[is_nan] = 0
    return v.sum(*args, **kwargs) / (~is_nan).float().sum(*args, **kwargs)
```
This is a part of the [pylabyk](https://github.com/yulkang/pylabyk) library, which also has `nansum`, among others. Please let me know if you find any issue.

**Comment by umanwizard** [comment]
I'm not working on this anymore, sorry.

**Comment by Novak3** [comment]
I came across this page by looking for a solution to the problem [referenced here:](https://discuss.pytorch.org/t/masked-batch-normalization/57813) namely, trying to mask tensors for various styles of normalization.

It seems like extending _exclude_nans_ into the mean and var functions would be the start of that, and then possibly as simple as extending those flags into fooNormxD functions.

Is it appropriate to open a new issue to request that full suite of functionality, or to leave this comment here?

**Comment by ezyang** [comment]
I think that's similar enough to this issue that it's fine here. I updated the issue description.

**Comment by henrique** [comment]
> Doesn't this do the job:
> 
> ```python
> def nanmean(v, *args, inplace=False, **kwargs):
>     if not inplace:
>         v = v.clone()
>     is_nan = torch.isnan(v)
>     v[is_nan] = 0
>     return v.sum(*args, **kwargs) / (~is_nan).float().sum(*args, **kwargs)
> ```
> 
> This is a part of my [pylabyk](https://github.com/yulkang/pylabyk) library, which also has `nansum`, among others. Please let me know if you find any issue.

@yulkang Isn't it the same as x[~torch.isnan(x)].mean()
Or am I missing some corner case?

**Comment by Novak3** [comment]
@henrique The method by @yulkang gracefully handles mean along axes. 
Maybe I'm missing something but I don't think your proposed method does-- it flattens the tensor

**Comment by henrique** [comment]
@Novak3 good point, you can use e.g. mean(dim=1) etc on his neat function.
It was killing my gradients, but it was perhaps just because of the `v.clone()` as I didn't see it had an inplace option.
Either way, if you don't care about the axes `x[~torch.isnan(x)].mean()` does the job slightly faster.

**Comment by jinserk** [comment]
The method by @yulkang has an issue if `(~is_nan).float().sum` part gives 0. That means the overall columns or rows are nan. Could it be workaround?

**Comment by ShirAmir** [comment]
Anything new regarding this feature?

**Comment by heitorschueroff** [comment]
> Anything new regarding this feature?

Yes, it's in our roadmap to implement the missing nan* ops.

**Comment by yulkang** [comment]
> The method by @yulkang has an issue if `(~is_nan).float().sum` part gives 0. That means the overall columns or rows are nan. Could it be workaround?

This is the default behavior for np.nanmean(). That said, I added an option to set `allnan` to a different value (e.g., 0):

```python
def nanmean(v: torch.Tensor, *args, allnan=np.nan, **kwargs) -> torch.Tensor:
    """

    :param v: tensor to take mean
    :param dim: dimension(s) over which to take the mean
    :param allnan: value to use in case all values averaged are NaN.
        Defaults to np.nan, consistent with np.nanmean.
    :return: mean.
    """
    v = v.clone()
    is_nan = isnan(v)
    v[is_nan] = 0

    if np.isnan(allnan):
        return v.sum(*args, **kwargs) / float(~is_nan).sum(*args, **kwargs)
    else:
        sum_nonnan = v.sum(*args, **kwargs)
        n_nonnan = float(~is_nan).sum(*args, **kwargs)
        mean_nonnan = torch.zeros_like(sum_nonnan) + allnan
        any_nonnan = n_nonnan > 1
        mean_nonnan[any_nonnan] = (
                sum_nonnan[any_nonnan] / n_nonnan[any_nonnan])
        return mean_nonnan
```

This is a part of my [pylabyk](https://github.com/yulkang/pylabyk) library, which also has `nansum`, among others. Please let me know if you find any issue.

**Comment by yulkang** [comment]
For those who wonder about related efforts:

* There seems to be an active PR: https://github.com/pytorch/pytorch/pull/38632
* And a rollup of numpy-like feature requests: https://github.com/pytorch/pytorch/issues/38349

**Comment by mruberry** [comment]
> For those who wonder about related efforts:
> 
> * There seems to be an active PR: #38632
> * And a rollup of numpy-like feature requests: #38349

And @heitorschueroff plans to work on nan* operations soon, so there's a good chance this will be appear in the next PyTorch release.

**Comment by addisonklinke** [comment]
A `torch.nanmean` option would be really useful as a reduction function for the `self.log(metric, on_epoch=True, sync_dist=True)` method in [PyTorch Lightning](https://github.com/PyTorchLightning/pytorch-lightning). Using `np.nanmean` is not an option for multi-GPU setups because their DDP accelerator requires `torch.Tensor` types, and Numpy can't handle those if they're on GPU memory

**Comment by mitchellgoffpc** [comment]
Any progress on this? This would be an extremely useful function!

**Comment by mruberry** [comment]
Unfortunately this didn't make the 1.9 release, but our plan is to implement it very soon, @mitchellgoffpc

**Comment by heitorschueroff** [comment]
Closed by https://github.com/pytorch/pytorch/pull/62671

**Comment by Ritsch1** [comment]
If someone still looks for a workaround:
